File: android/src/main/kotlin/com/example/microsoft_clarity_plugin/MicrosoftClarityPlugin.kt
package com.example.microsoft_clarity_plugin

import android.annotation.SuppressLint
import android.content.Context
import com.microsoft.clarity.Clarity
import com.microsoft.clarity.ClarityConfig
import com.microsoft.clarity.models.ApplicationFramework
import com.microsoft.clarity.models.LogLevel

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** MicrosoftClarityPlugin */
class MicrosoftClarityPlugin: FlutterPlugin, MethodCallHandler {
  private lateinit var channel: MethodChannel
  private lateinit var context: Context

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    context = flutterPluginBinding.applicationContext
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "microsoft_clarity_plugin")
    channel.setMethodCallHandler(this)
  }

  @SuppressLint("DiscouragedApi")
  override fun onMethodCall(call: MethodCall, result: Result) {
    if (call.method == "initializeMicrosoftClarity") {
      val resId = context.resources.getIdentifier("microsoft_clarity_project_id", "string", context.packageName)
      if (resId != 0) {
        val projectId = context.getString(resId)
        initializeMicrosoftClarity(context, projectId)
        result.success("Clarity Initialized")
      } else {
        result.error("NOT_FOUND", "Project ID not found", null)
      }
    } else if(call.method == "setCustomUserId") {
      val customUserId: String? = call.argument("customUserId");
      if (customUserId == null) {
        result.error("INVALID_ARGUMENT", "customUserId is required", null)
      } else {
        Clarity.setCustomUserId(customUserId)
        result.success("Custom User Id set successfully")
      }
    } else if(call.method == "getCurrentSessionId") {
      result.success(Clarity.getCurrentSessionId());
    } else {
      result.notImplemented()
    }
  }

  private fun initializeMicrosoftClarity(context: Context?, projectId: String) {
    context?.let {
      val config = ClarityConfig(
        projectId = projectId,
        userId = null, // Default user id
        logLevel = LogLevel.None,
        allowMeteredNetworkUsage = true, // Disallow metered network usage
        enableWebViewCapture = true, // Enable web view capturing
        allowedDomains = listOf("*"), // Allowed domains
        applicationFramework = ApplicationFramework.Native,
        allowedActivities = listOf(), // Allowed activities
        disallowedActivities = listOf(), // Disallowed activities (ignore activities)
        disableOnLowEndDevices = false,
        maximumDailyNetworkUsageInMB = null,// Disable on low-end devices
      )
      Clarity.initialize(it.applicationContext, config)
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}

File: ios/Classes/MicrosoftClarityPlugin.swift
import Flutter
import UIKit
import Clarity

public class MicrosoftClarityPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "microsoft_clarity_plugin", binaryMessenger: registrar.messenger())
    let instance = MicrosoftClarityPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
      case "initializeMicrosoftClarity":
        if let projectId = Bundle.main.object(forInfoDictionaryKey: "MICROSOFT_CLARITY_PROJECT_ID") as? String {
          initializeMicrosoftClarity(projectId: projectId)
          print("MICROSOFT_CLARITY_PROJECT initialized successfully.")
          result("clarity-project-initialization-success")
        } else {
          print("MICROSOFT_CLARITY_PROJECT_ID not found in Info.plist file.")
          result(FlutterError(code: "MISSING_MICROSOFT_CLARITY_PROJECT_ID", message: "project id is required", details: nil))
        }
      case "setCustomUserId":
        if let args = call.arguments as? Dictionary<String, Any>, let id = args["customUserId"] as? String {
          ClaritySDK.setCustomUserId(id)
          print("CUSTOM_USER_ID was not provided.")
          result("custom-user-id-setting-success")
        } else {
          print("CUSTOM_USER_ID not passed in arguments")
          result(FlutterError(code: "MISSING_CUSTOM_USER_ID", message: "custom user ID is required", details: nil))
        }
      case "getCurrentSessionId":
        result(ClaritySDK.getCurrentSessionId())
      default:
        result(FlutterMethodNotImplemented)
    }
  }

  private func initializeMicrosoftClarity(projectId: String) {
    let clarityConfig = ClarityConfig(
      projectId: projectId,
      userId: nil, // Default user id
      logLevel: LogLevel.verbose,
      allowMeteredNetworkUsage: true, // Allow metered network usage
      enableWebViewCapture: true, // Enable web view capturing
      disableOnLowEndDevices: false, // Allow running on low-end devices
      applicationFramework: ApplicationFramework.native
    )
    ClaritySDK.initialize(config: clarityConfig)
  }
}